src/png.rs

```diff
@@ -218,26 +250,68 @@ impl<W: Write> PNGEncoder<W> {
+    /// with ```CompressionType``` ```compression``` and
+    /// ```FilterType``` ```filter```.
+    pub fn encode_with_settings(
+        self,
+        data: &[u8],
+        width: u32,
+        height: u32,
+        color: ColorType,
+        compression: CompressionType,
+        filter: FilterType,
+    ) -> ImageResult<()> {
         let (ct, bits) = match color {
             ColorType::L8 => (png::ColorType::Grayscale, png::BitDepth::Eight),
-            ColorType::L16 => (png::ColorType::Grayscale,png::BitDepth::Sixteen),
+            ColorType::L16 => (png::ColorType::Grayscale, png::BitDepth::Sixteen),
             ColorType::La8 => (png::ColorType::GrayscaleAlpha, png::BitDepth::Eight),
-            ColorType::La16 => (png::ColorType::GrayscaleAlpha,png::BitDepth::Sixteen),
+            ColorType::La16 => (png::ColorType::GrayscaleAlpha, png::BitDepth::Sixteen),
             ColorType::Rgb8 => (png::ColorType::RGB, png::BitDepth::Eight),
-            ColorType::Rgb16 => (png::ColorType::RGB,png::BitDepth::Sixteen),
+            ColorType::Rgb16 => (png::ColorType::RGB, png::BitDepth::Sixteen),
             ColorType::Rgba8 => (png::ColorType::RGBA, png::BitDepth::Eight),
-            ColorType::Rgba16 => (png::ColorType::RGBA,png::BitDepth::Sixteen),
-            _ => return Err(ImageError::Unsupported(UnsupportedError::from_format_and_kind(
-                ImageFormat::Png.into(),
-                UnsupportedErrorKind::Color(color.into()),
-            ))),
+            ColorType::Rgba16 => (png::ColorType::RGBA, png::BitDepth::Sixteen),
+            _ => {
+                return Err(ImageError::Unsupported(
+                    UnsupportedError::from_format_and_kind(
+                        ImageFormat::Png.into(),
+                        UnsupportedErrorKind::Color(color.into()),
+                    ),
+                ))
+            }
+        };
+
+        let c = match compression {
+            CompressionType::Default => png::Compression::Default,
+            CompressionType::Fast => png::Compression::Fast,
+            CompressionType::Best => png::Compression::Best,
+            CompressionType::Huffman => png::Compression::Huffman,
+            CompressionType::Rle => png::Compression::Rle,
+        };
+        let f = match filter {
+            FilterType::NoFilter => png::FilterType::NoFilter,
+            FilterType::Sub => png::FilterType::Sub,
+            FilterType::Up => png::FilterType::Up,
+            FilterType::Avg => png::FilterType::Avg,
+            FilterType::Paeth => png::FilterType::Paeth,
         };
 
         let mut encoder = png::Encoder::new(self.w, width, height);
         encoder.set_color(ct);
         encoder.set_depth(bits);
-        let mut writer = encoder.write_header().map_err(|e| ImageError::IoError(e.into()))?;
-        writer.write_image_data(data).map_err(|e| ImageError::IoError(e.into()))
+        encoder.set_compression(c);
+        encoder.set_filter(f);
```

Comment:
I don't remember but I don't fully understand your question.

From what I can remember, the color type and bit depth were already being set there so I added the compression and filter type. It seemed like each parameter had to be set individually on the encoder. The actual png::Encoder isn't created until you call `encode`, before that it's wrapping a writer. This didn't change the signature of `encode` similar to how making a new JPEGEncoder calls `new_with_quality` from `new`.

The other plan I thought of but avoided because of being a breaking change was to add fields to PngEncoder like JpegEncoder. Then, there could be a PngEncoder `new_with_*` function where the encoder settings are set, like compression and filter type. The struct encoder settings would be read by the `encode` method without changing the argument list.

Reply:
> The other plan I thought of but avoided because of being a breaking change was to add fields to PngEncoder like JpegEncoder.

 I'm not sure I follow. How would that be a breaking change to add fields? The struct is currently private and we already have default values for those fields that can be used in the other constructors.

Reply:
I see that now, I misread the code.